sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -2021,88 +2032,218 @@ class Analyzer(override val catalogManager: CatalogManager)
+      case other if ignoreNulls =>
+        throw QueryCompilationErrors.functionWithUnsupportedSyntaxError(
+          other.prettyName, "IGNORE NULLS")
+      case e: String2TrimExpression if arguments.size == 2 =>
+        if (trimWarningEnabled.get) {
+          log.warn("Two-parameter TRIM/LTRIM/RTRIM function signatures are deprecated." +
+            " Use SQL syntax `TRIM((BOTH | LEADING | TRAILING)? trimStr FROM str)`" +
+            " instead.")
+          trimWarningEnabled.set(false)
+        }
+        e
+      case other =>
+        other
+    }
+
+    private def lookupV2Function(
+        catalog: FunctionCatalog,
+        ident: Identifier,
+        arguments: Seq[Expression],
+        isDistinct: Boolean,
+        filter: Option[Expression],
+        ignoreNulls: Boolean): Expression = {
+      val unbound = catalog.loadFunction(ident)
+      val inputType = StructType(arguments.zipWithIndex.map {
+        case (exp, pos) => StructField(s"_$pos", exp.dataType, exp.nullable)
+      })
+      val bound = try {
+        unbound.bind(inputType)
+      } catch {
+        case unsupported: UnsupportedOperationException =>
+          failAnalysis(s"Function '${unbound.name}' cannot process input: " +
+            s"(${arguments.map(_.dataType.simpleString).mkString(", ")}): " +
+            unsupported.getMessage)
+      }
+
+      bound match {
+        case scalarFunc: ScalarFunction[_] =>
+          if (isDistinct) {
+            throw QueryCompilationErrors.functionWithUnsupportedSyntaxError(
+              scalarFunc.name(), "DISTINCT")
+          } else if (filter.isDefined) {
+            throw QueryCompilationErrors.functionWithUnsupportedSyntaxError(
+              scalarFunc.name(), "FILTER clause")
+          } else if (ignoreNulls) {
+            throw QueryCompilationErrors.functionWithUnsupportedSyntaxError(
+              scalarFunc.name(), "IGNORE NULLS")
+          } else {
+            // TODO: implement type coercion by looking at input type from the UDF. We may
+            //  also want to check if the parameter types from the magic method match the
+            //  input type through `BoundFunction.inputTypes`.
+            val argClasses = inputType.fields.map(_.dataType)
+            findMethod(scalarFunc, MAGIC_METHOD_NAME, Some(argClasses)) match {
+              case Some(_) =>
+                val caller = Literal.create(scalarFunc, ObjectType(scalarFunc.getClass))
+                Invoke(caller, MAGIC_METHOD_NAME, scalarFunc.resultType(),
+                  arguments, returnNullable = scalarFunc.isResultNullable)
+              case _ =>
+                // TODO: handle functions defined in Scala too - in Scala, even if a
+                //  subclass do not override the default method in parent interface
+                //  defined in Java, the method can still be found from
+                //  `getDeclaredMethod`.
+                findMethod(scalarFunc, "produceResult", Some(Seq(inputType))) match {
```

Comment:
Isn't this supposed to look up the method that accepts `InternalRow`? But it doesn't look like that's what happens here because the sequence of input types is passed.